cmd/metal-api/internal/service/async-actor.go

```diff
@@ -0,0 +1,167 @@
+	if err != nil {
+		return err
+	}
+
+	// call the releaser async
+	err = a.machineReleaser(m)
+	if err != nil {
+		// log error, but what should we do here? we already called
+		// deleteVRFSwitches and publishDeleteEvent, so should we return
+		// an error or "fall through"?
+		a.Error("cannot call async machine cleanup", zap.Error(err))
+	}
+
+	old := *m
+
+	m.Allocation = nil
+	m.Tags = nil
+
+	err = a.UpdateMachine(&old, m)
+	if err != nil {
+		return err
+	}
+	a.Info("freed machine", zap.String("machineID", m.ID))
+
+	return nil
+}
+
+func (a *asyncActor) releaseMachineNetworks(machine *metal.Machine) error {
+	if machine.Allocation == nil {
+		return nil
+	}
+	for _, machineNetwork := range machine.Allocation.MachineNetworks {
+		for _, ipString := range machineNetwork.IPs {
+			ip, err := a.FindIPByID(ipString)
+			if err != nil {
+				return err
+			}
+			// ignore ips that were associated with the machine for allocation but the association is not present anymore at the ip
+			if !ip.HasMachineId(machine.GetID()) {
+				continue
+			}
+			// disassociate machine from ip
+			new := *ip
+			new.RemoveMachineId(machine.GetID())
+			err = a.UpdateIP(ip, &new)
+			if err != nil {
+				return err
+			}
+			// static ips should not be released automatically
+			if ip.Type == metal.Static {
+				continue
+			}
+			// ips that are associated to other machines will should not be released automatically
+			if len(new.GetMachineIds()) > 0 {
+				continue
+			}
+
+			// at this point the machine is removed from the IP, so the release of the
+			// IP must not fail, because this loop will never come to this point again because the
+			// begin of this loop checks if the IP contains the machine.
+			// so we fork a new job to delete the IP. if this fails .... well then ("houston we have a problem")
+			// we do not report the error to the caller, because this whole function cannot be re-do'ed.
+			a.Infow("async release IP", "ip", *ip)
+			if err := a.ipReleaser(*ip); err != nil {
+				// what should we do here? this error shows a problem with the nsq-bus system
+				a.Error("cannot call ip releaser", zap.Error(err))
+			}
+		}
+	}
+	return nil
+}
+
+func (a *asyncActor) releaseIP(ip metal.IP) error {
+	a.Info("release IP", zap.Any("ip", ip))
+
+	dbip, err := a.FindIPByID(ip.IPAddress)
+	if err != nil && !metal.IsNotFound(err) {
+		// some unknown error, we will let nsq resend the command
+		a.Error("cannot find IP", zap.Any("ip", ip), zap.Error(err))
+		return err
+	}
+
+	if err == nil {
+		// if someone calls freeMachine for the same machine multiple times at the same
+		// moment it can happen that we already deleted and released the IP in the ipam
+		// so make sure that this IP is not already connected to a new machine
+		if len(dbip.GetMachineIds()) > 0 {
+			a.Info("do not delete IP, it is connected to a machine", zap.Any("ip", ip))
+			return nil
+		}
+		s := dbip.GetScope()
+		if s != metal.ScopeProject && ip.Type == metal.Static {
```

Comment:
I don't not really get this condition. The log message says something that this condition does not express. I am not sure what should actually be prevented here.

Can there even be static IPs on another scope than Project Scope?

Reply:
i only moved the code from here: https://github.com/metal-stack/metal-api/blob/4a2153efe97d37ebffbc8b419b4556c08095b3b8/cmd/metal-api/internal/service/ip-service.go#L183

to the actor. about two weeks ago we discussed this validation logic, because i also was a little puzzled about this one. it was created by @mwindower in this commit https://github.com/metal-stack/metal-api/commit/472e4aa1a3f014068ddb914185ed11a9a1bcc4ae

Reply:
@mwindower Can you please provide your opinion to this? Is this validation required for IP cleanup on `machineFree` as well or only when a user calls IP delete?

Reply:
This is about static IPs that were directly assigned to machines. If the scope is `ScopeMachine` this means that there is at least one machine that uses it. In that case `freeIP` should fail. 
`freeMachine` or the actor should only delete the machine tag from those IPs but never delete it.

Reply:
Ok. I think this information helps me a bit. Let's go through two scenarios:

Scenario 1:

- A user allocates a static IP address
- A user allocates a machine and adds the IP via `--ips` 
- A user calls machine free 
  - The IP should not be released 
  - The machine tag should be removed from the IP

Scenario 2 (I hope this is correct):

- A user allocates a static IP address
- A user allocates a machine and adds the IP via `--ips` 
- A user calls ip free on the IP address
  - The IP should not be released 
  - The IP machine tag should not be removed because there is still a machine depending on it (otherwise we have a machine with an IP in the machine allocation that the machine does not own anymore)
  - This actually means that nothing happened

Therefore I guess that we should actually check if a machine still depends on an IP address in `freeIP` before handing it over to the reconcile. Just raise an error and let the user free the machines first before he can free the static ip. With the current implementation it would look to the user like `freeIP` was a success but actually nothing happened except our logger said "do not delete static IP" but unfortunately the user still owns the IP without knowing why. 

I am not sure if I made a mistake in my reasoning though. But if this is correct, the check does not belong in the reconcile but only in `freeIP`.

Reply:
When I think about this, I would assume that it even does not matter in this case if an IP is static or ephemeral. The error message should be something like "cannot free ip address with machines depending on it". And it should always be checked in `freeIP`.

Reply:
Yes, I agree:
- always check for machines depending on an IP (ephemeral as well as static in `freeIP` and `freeMachine`)
- let `freeMachine` only delete ephemeral IPs

Reply:
@mwindower is the condition 
~~~
if s != metal.ScopeProject && ip.Type == metal.Static {
~~~
identical to
~~~
if s == metal.ScopeMachine
~~~
?

this would be much more intuitive.

Reply:
As @Gerrit91 mentioned: the whole condition is unnecessary. We could just remove the whole if-block `if s != metal.ScopeProject && ip.Type == metal.Static`